ldap/servers/slapd/mapping_tree.c

```diff
@@ -2058,6 +2058,82 @@ slapi_dn_write_needs_referral(Slapi_DN *target_sdn, Slapi_Entry **referral)
 done:
     return ret;
 }
+
+/*
+ * This function dermines if an operation should be rejected
+ * when readonly mode is enabled.
+ * All operations are rejected except:
+ *    - if they target a private backend that is not the DSE backend
+ *    - if they are read operations (SEARCH, COMPARE, BIND, UNBIND)
+ *    - if they are tombstone fixup operation (i.e: tombstone purging)
+ *    - if they are internal operation that targets the DSE backend.
+ *      (change will then be done in memory but not written in dse.ldif)
+ *    - single modify modify operation on cn=config changing nsslapd-readonly
+ *      (to allow  "dsconf instance config replace nsslapd-readonly=xxx",
+         change will then be done both in memory and in dse.ldif)
+ */
+static bool
+is_rejected_op(Slapi_Operation *op, Slapi_Backend *be)
+{
+    const char *betype = slapi_be_gettype(be);
+    unsigned long be_op_type = operation_get_type(op);
+    int isdse = (betype && strcmp(betype, "DSE") == 0);
+
+    /* Private backend operations are not rejected */
+
+    /* Read operations are not rejected */
+    if ((be_op_type == SLAPI_OPERATION_SEARCH) ||
+        (be_op_type == SLAPI_OPERATION_COMPARE) ||
+        (be_op_type == SLAPI_OPERATION_BIND) ||
+        (be_op_type == SLAPI_OPERATION_UNBIND)) {
+        return false;
+    }
+
+    /* Tombstone fixup are not rejected. */
+    if (operation_is_flag_set(op, OP_FLAG_TOMBSTONE_FIXUP)) {
```

Comment:
Should we also accept RUV update (OP_FLAG_REPL_RUV) ? As there are no update, RUV should not change so it is probably fine to ignore RUV.
Or it may be handle by later test that accept all internal operation.